### ISSUE TITLE
模型训练显存占用与max_seq_length问题

### ISSUE THREAD

**Issue by Aurora-6** [hiyouga/ChatGLM-Efficient-Tuning#111]
使用项目代码commit 2dbba2f7942f046ae1eb113b1aff955510b7a530，进行reward训练，GPUs: A100(40G)*8
reward模型训练：max_source_length=256, max_target_length=256训到400+samples就OOM了，max_length只能更小
换成多机多卡也没有能扩充max_length

想问下正常的训练情况是什么能力呢？

**Comment by hiyouga** [comment]
batch size 是多大？用 deepspeed 进行训练呢？
确实有人反馈过多卡容易 OOM 的问题，似乎单卡要更稳定。

**Comment by Aurora-6** [comment]
> batch size 是多大？用 deepspeed 进行训练呢？
> 确实有人反馈过多卡容易 OOM 的问题，似乎单卡要更稳定。

batch_size=1...  用的是accelerate

**Comment by Aurora-6** [comment]
max_length调小到128之后，也能看到显存富裕着，但调大训着训着就OOM了。尝试了一下del变量，empty_cache这些有用处但不大。

**Comment by hiyouga** [comment]
单卡运行呢？

**Comment by hiyouga** [comment]
我这边也是用 A100 40G 调试的，三个脚本都可以开到 batch size=8。其余参数均和默认一致，检查一下是不是 CUDA 环境或者版本问题。

**Comment by hiyouga** [comment]
单卡A100，运行的都是 examples 里面的脚本

**Comment by Aurora-6** [comment]
> 单卡A100，运行的都是 examples 里面的脚本

好的，我再检查下，谢谢～

**Comment by Aurora-6** [comment]
我用了项目中reward example训练，都遵循默认配置，data也用的是comparison_gpt4_zh，max_length两个都512可以正常训练。但是换成我自己的数据集，prompt都是偏长的数据（500+字），512完全不行，也就是昨天说的只能256以下。

我在想之所以可以设成512是不是因为例子的训练数据基本到不了512tokens所以没有OOM呢？

**Comment by Aurora-6** [comment]
好的。因为代码里是对超出max_length的数据直接截断，所以我就直接用几条1000+字的prompt数据训练模型，batch_size=1，不断调整max_length大小，实验发现8*A100(40G)训练能力：
PPO训练：max_source_length=256能正常训练，此时显存已经占用38900MiB。PPO这步显存占用的应该就只是任务通用的6b模型（13G+）, set, reward, rlhf的三个lora小模型。为什么显存占用这么大呢？

**Comment by Aurora-6** [comment]
另外在看ppo代码的时候有个问题想请教下，train_ppo.py里为什么ppo_trainer设置里ref_model=None呢？
`# Initialize our Trainer
    ppo_trainer = PPOTrainerForChatGLM(
        training_args=training_args,
        finetuning_args=finetuning_args,
        config=ppo_config,
        model=model,
        ref_model=None,
        tokenizer=tokenizer,
        dataset=dataset,
        data_collator=data_collator,
        optimizer=optimizer,
        lr_scheduler=lr_scheduler
    )`

**Comment by Aurora-6** [comment]
> 另外在看ppo代码的时候有个问题想请教下，train_ppo.py里为什么ppo_trainer设置里ref_model=None呢？
> 
> ```
> # Initialize our Trainer
>     ppo_trainer = PPOTrainerForChatGLM(
>         training_args=training_args,
>         finetuning_args=finetuning_args,
>         config=ppo_config,
>         model=model,
>         ref_model=None,
>         tokenizer=tokenizer,
>         dataset=dataset,
>         data_collator=data_collator,
>         optimizer=optimizer,
>         lr_scheduler=lr_scheduler
>     )
> ```

看了下trl库函数的解释理解啦，
```
**ref_model** (`PreTrainedModelWrapper`, *optional*) -- Reference model to be used for KL penalty, Hugging Face
            transformer model with a casual language modelling head. Check the documentation of `PreTrainedModelWrapper`
            for more details. If no reference model is provided, the trainer will create a reference model with the same
             architecture as the model to be optimized with shared layers.
```

**Comment by jiahuanluo** [comment]
> 好的。因为代码里是对超出max_length的数据直接截断，所以我就直接用几条1000+字的prompt数据训练模型，batch_size=1，不断调整max_length大小，实验发现8*A100(40G)训练能力： PPO训练：max_source_length=256能正常训练，此时显存已经占用38900MiB。PPO这步显存占用的应该就只是任务通用的6b模型（13G+）, set, reward, rlhf的三个lora小模型。为什么显存占用这么大呢？

请问后来解决了吗？我单卡V100-32G跑例子可以跑通，换到自己的比较长的数据，跑ppo就oom了。不知道能不能通过多卡解决oom的问题呢？
